lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/MetricsSystem.cs

```
@@ -0,0 +1,307 @@
+        /// sink handler.
+        /// </summary>
+        /// <param name="all">Whether to also publish metrics that did not change.</param>
+        public void PublishMetricsNow(bool all)
+        {
+            lock (_lock)
+            {
+                if (_shutDown)
+                {
+                    return;
+                }
+                PublishMetrics(all);
+            }
+        }
+
+        /// <summary>
+        /// Shuts down the metrics system. Call Stop() befor ShutDown()
+        /// </summary>
+        /// <returns>True if shutdown is successful.</returns>
+        public bool ShutDown()
+        {
+            lock (_lock)
+            {
+                _shutDown = true;
+                foreach (var sink in _sinks)
+                {
+                    sink.Value.ShutDown();
+                }
+                _sinks.Clear();
+                _sources.Clear();
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Function called to take snapshot from sources and put the in the sink queue. 
+        /// This method is either called periodically or by user via. <see cref="PublishMetricsNow"/>.
+        /// </summary>
+        private void PublishMetrics(bool all)
+        {
+            lock (_lock)
+            {
+                if (_sinks.Count != 0 && _sources.Count != 0)
+                {
+                    _sourceSnapshots.Clear();
+                    foreach (var source in _sources)
+                    {
+                        if (_sourceFilter.AcceptsName(source.Key))
+                        {
+                            _metricsCollector.Clear();
+                            _sourceSnapshots.Add(source.Value.GetMetrics(_metricsCollector, all));
+                        }
+                    }
+                    foreach (var sink in _sinks)
+                    {
+                        sink.Value.PutMetricsInQueue(_sourceSnapshots);
```

I don't think it's a good idea to just silently discard metrics if queue is full. Let's at least add a log, so that we know that these metrics are discarded intentionally